ФЗ-12. Add database

Black-Japan-Tea opened the pull request.

Black-Japan-Tea committed:
Фз-12. add-database. init

Black-Japan-Tea committed:
Delete checkstyle.xml

Black-Japan-Tea committed:
Фз-12. add-database. fix

Black-Japan-Tea committed:
Merge remote-tracking branch 'origin/add-database' into add-database

Black-Japan-Tea committed:
Фз-12. add-database. fix

Black-Japan-Tea committed:
Работа с БД.

Black-Japan-Tea committed:
add checkstyle

Black-Japan-Tea committed:
Revert checkstyle.xml to original

Black-Japan-Tea committed:
fix schema.erd

Black-Japan-Tea committed:
add checkstyle

Black-Japan-Tea committed:
Merge remote-tracking branch 'origin/add-database' into add-database

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Optional;
+
+@Service
+public class GenreService {
```
добавь в класс логирование

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/MpaRateService.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Optional;
+
+@Service
+public class MpaRateService {
```
в этот класс тоже нужно добавить логирование

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,105 +1,151 @@
 import java.util.Optional;
-import java.util.stream.Collectors;
 
 @Slf4j
```
в этом классе их тоже маловато

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,115 +1,172 @@
+import java.util.Set;
 import java.util.stream.Collectors;
 
 @Slf4j
```
как и в этом

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,115 +1,172 @@
+        if (newUserId == 0) {
+            throw new StorageException(newUser + " wasn't created");
+        }
+        Optional<User> userInStorage = userStorage.findUserById(newUserId);
```
воспользуйся checkAndGetUserById()

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,115 +1,172 @@
-
-        user.getFriends().remove(friendId);
-        friend.getFriends().remove(userId);
+        if (userId == friendId) {
```
сделай эту проверку первой в методе. Если сразу понятно, что на вход пришли не валидные данные, к чему нагружать БД бесполезными запросами

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,115 +1,172 @@
 
-        userStorage.updateUser(user);
-        userStorage.updateUser(friend);
+        if (isUserHaveFriend(userId, friendId)) {
```
давай еще немного оптимизируем процесс. Немного выше запиши в переменную результат checkAndGetUserById(userId) и список друзей возьми из переменной. А метод isUserHaveFriend() сделай вот таким isUserHaveFriend(Set<User> getUserFriends, long friendId)

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,115 +1,172 @@
+    public void deleteFriend(long userId, long friendId) {
+        checkAndGetUserById(userId);
+        checkAndGetUserById(friendId);
+        if (userId == friendId) {
```
тоже в самое начало метода

irek-valiev commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+
+    private void addCollectionsToFilms(Collection<Film> films) {
```
не нужно в цикле обращаться к БД за данными, каждое соединение с БД - это затратная процедура. Заранее вытяни из БД все данные жанрам и фильмам, в потоке из них собери мапу типа Map<film.getId(), List Жанры>, а далее уже бери жанры из этой мапы. Аналогично по лайкам

Black-Japan-Tea committed:
Правки согласно ревью:
Добавлено логирование сервисов
Работа с проверкой данных на валидность
Улучшено взаимодействие с БД
